src/buffer_manager.ts

```diff
@@ -622,8 +624,39 @@ export class BufferManager implements Disposable {
             this.textDocumentToBufferId.delete(document);
             unusedBuffers.push(bufId);
         });
-        unusedWindows.length && (await actions.lua("close_windows", unusedWindows));
-        unusedBuffers.length && (await actions.lua("delete_buffers", unusedBuffers));
+
+        logger.debug(`Closing windows: ${inspect(unusedWindows)}...`);
+        for (const winId of unusedWindows) {
+            logger.debug(`Closing window ${winId}...`);
+            try {
+                await this.request("nvim_win_close", [winId, true], 5000);
```

Comment:
The Lua code runs in the Nvim server. With your changes, now multiple rpc requests would be made (one request per window).

Reply:
Ah, I guess that makes sense as a reason to reduce the overhead of the call. I guess one downside of batching like that would be that it is harder to tell _which_ call caused the timeout, unless the timeout mechanism itself were implemented in Lua and e.g. a list of results was returned...

Reply:
> harder to tell _which_ call caused the timeout

well, `nvim_win_close` cannot hang or timeout. so the only thing that could be timing out is the RPC call.

Reply:
Hmm, I thought that when I had added prints on the Lua side (before bypassing it) that even the `pcall(vim.api.nvim_win_close, win, true)` was getting stuck, but it was a while ago and I don't know much about the nvim internals that are doing this work... 

Assuming it is the RPC, does that imply we should anticipate a possible hang like this on _any_ call and write the client to handle those cases? Would that be a neovim core bug or a `node-client` bug?

Reply:
> does that imply we should anticipate a possible hang like this on _any_ call and write the client to handle those cases?

consumers like vscode-neovim shouldn't generally worry about that. I think https://github.com/neovim/node-client/ or Nvim should eventually try to provide a generalized timeout/watchdog mechanism.